### PR TITLE
Ipc groundwork

### PR DESCRIPTION
**Pull request by ramfox**
This pr lays the groundwork for a streaming rpc protocol.

It includes implementations of a database and cli client, and a database "provider" that responds to rpc requests.

Use `cargo run --bin iroh-rpc -- --file FILE_PATH --transport [mem|tcp]` to test streaming a file from one processes (the database process) to another (the cli process).

The "streaming" protocol built on top of the libp2p `request_response` protocol, using a `DataRequest` and a `Header` response to coordinate a streaming `id`. The database provider then sends `Packets` notated with the proper `id` using the "streaming" protocol.

All of the streaming coordination happens via the `InStream` and `OutStream` structs.

`InStream::new` takes a header & a receiver of a buffered channel, and returns an `InStream`. Calling `stream.next().await` gives you the next chunk of bytes for the expected file, in the correct order.

`OutStream::new` take a header, a reader, and a sender, and returns an `OutStream`. Calling `stream.send_packets().await`  sends the packets over rpc. 

Future refactors should:
- allow for early cancel on both sides of the connection
- handle `Ack` and `Error` responses from the sink to the source. Potentially re-sending packets if an `Error` has occurred
- handle "channel full" errors on the sink side, sending an error back to the source. The source should back off for a few moments before attempting to send again.
- handle reading errors on the source side, informing the sink

### CODE REVIEW

File: iroh-rpc/src/core.rs [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+use libp2p::noise;
+use libp2p::yamux;
+
+pub fn mem_transport(keypair: Keypair) -> transport::Boxed<(PeerId, muxing::StreamMuxerBox)> {
+    let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+        .into_authentic(&keypair)
+        .unwrap();
+    MemoryTransport::default()
+        .upgrade(upgrade::Version::V1)
+        .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
```
Comment (dignifiedquire) [inline iroh-rpc/src/core.rs:433]: no need for encryption on the memory transport

Comment (dignifiedquire) [reply iroh-rpc/src/core.rs:433]: if you switch to plaintext it shaves off another ~20%

Comment (dignifiedquire) [reply iroh-rpc/src/core.rs:433]: https://docs.rs/libp2p/latest/libp2p/plaintext/struct.PlainText2Config.html

Comment (ramfox) [reply iroh-rpc/src/core.rs:447]: omg rock star thank you.

File: iroh-rpc/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            db_in_receiver,
+            db_addr,
+        }
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    {
+        let matches = Command::new("iroh-ipc")
```
Comment (dignifiedquire) [inline iroh-rpc/src/main.rs:78]: this should be configured through a struct (much nicer to maintain and handle)

File: iroh-rpc/src/core.rs [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+
+/// Build a swarm with a TCP transport
+// TODO: eventually generalize, pass in config etc
+pub async fn new_swarm(id_keys: Keypair) -> Result<Swarm<CoreBehaviour>, Box<dyn Error>> {
+    let peer_id = id_keys.public().to_peer_id();
+    Ok(SwarmBuilder::new(
+        libp2p::development_transport(id_keys).await?,
+        CoreBehaviour::new(),
+        peer_id,
+    )
```
Comment (dignifiedquire) [inline iroh-rpc/src/core.rs:435]: this is missing setting up tokio as the executor, this is pretty important: https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L140-L144

File: iroh-rpc/src/core.rs [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+pub struct InboundEvent {
+    pub command: InCommand,
+    pub channel: StreamingResponseChannel,
+}
+
+/// Build a swarm with an in memory transport.
+// TODO: eventually generalize, pass in config etc
+pub fn new_mem_swarm(id_keys: Keypair) -> Swarm<CoreBehaviour> {
+    let peer_id = id_keys.public().to_peer_id();
+    SwarmBuilder::new(mem_transport(id_keys), CoreBehaviour::new(), peer_id).build()
```
Comment (dignifiedquire) [inline iroh-rpc/src/core.rs:424]: this is missing setting up tokio as the executor, this is pretty important: https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L140-L144

**Comment by ramfox** [comment]
This PR is no longer relevant, closing

**Pull request closed without merging**